### DISCUSSION TITLE
How to receive a list (indexed with squared brackets) of objects in query parameters?

### DISCUSSION THREAD

**Discussion by ldbolanos** [fastapi/fastapi#11399] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn as uvicorn
from fastapi import FastAPI, Path, Query, Depends
from pydantic import BaseModel

class SearchQuery(BaseModel):
    value: str | None = None
    regex: bool | None = None


class ColumnQuery(BaseModel):
    name: str | None = None
    searchable: bool | None = None
    orderable: bool | None = None
    search: SearchQuery = Depends()


class TableQuery(BaseModel):
    draw: int
    start: int
    length: int
    columns: ColumnQuery = Depends() # Only keeps last item from list
    # columns: list[ColumnQuery] = Depends() # Does not compile


app = FastAPI()

@app.get("/data")
def does_not_work(params: TableQuery = Depends()):
    print(params)

# draw=55 start=0 length=10 columns=ColumnQuery(name=None, searchable=None, orderable=None, search=SearchQuery(value=None, regex=None))
```

### Description

I'm working in the JS library [DataTables](https://datatables.net/) and want to implement the server-side processing where the frontend uses pagination to display segments of a table.

I'm struggling because the [query parameters](https://datatables.net/manual/server-side#Sent-parameters) sent by DataTables are convoluted. In particular, the `columns` parameter is a nested list
```
columns[0][name]=test&columns[0][searchable]=true&columns[0][orderable]=false
```
and I have no idea how to parse it as a query parameter.

I am providing the code I'm using right now. Unfortunately, it only keeps the last item in the list.

This is the request from the client side.

```
curl 'http://localhost:8000?draw=55&columns%5B0%5D%5Bdata%5D=0&columns%5B0%5D%5Bname%5D=&columns%5B0%5D%5Bsearchable%5D=true&columns%5B0%5D%5Borderable%5D=true&columns%5B0%5D%5Bsearch%5D%5Bvalue%5D=5724585510000104000&columns%5B0%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B1%5D%5Bdata%5D=1&columns%5B1%5D%5Bname%5D=&columns%5B1%5D%5Bsearchable%5D=true&columns%5B1%5D%5Borderable%5D=true&columns%5B1%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B1%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B2%5D%5Bdata%5D=2&columns%5B2%5D%5Bname%5D=&columns%5B2%5D%5Bsearchable%5D=true&columns%5B2%5D%5Borderable%5D=true&columns%5B2%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B2%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B3%5D%5Bdata%5D=3&columns%5B3%5D%5Bname%5D=&columns%5B3%5D%5Bsearchable%5D=true&columns%5B3%5D%5Borderable%5D=true&columns%5B3%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B3%5D%5Bsearch%5D%5Bregex%5D=false&order%5B0%5D%5Bcolumn%5D=0&order%5B0%5D%5Bdir%5D=asc&order%5B0%5D%5Bname%5D=&start=0&length=10&search%5Bvalue%5D=&search%5Bregex%5D=false&_=1712068863013' -H 'Accept: application/json, text/javascript, */*; q=0.01' -H 'Accept-Language: en-US,es-CO;q=0.7,en;q=0.3' -H 'Accept-Encoding: gzip, deflate, br' -H 'X-Requested-With: XMLHttpRequest' -H 'DNT: 1' -H 'Connection: keep-alive' -H 'Referer: http://localhost:8000/booking/' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-origin'
```

### Operating System

Windows

### FastAPI Version

0.103.1

### Pydantic Version

2.3.0

### Python Version

Python 3.10.13

**Comment by JavierSanchezCastro** [comment]
It's a bit complicated to handle that structure of data with Query parameters, you can use a POST to be able to use the body to send all this data (despite it's going to work as GET).

On the other hand, you can do this to send this structure of data by Query parameter: 

```python3
from fastapi import FastAPI, Query, HTTPException, status
from pydantic import BaseModel
import json

class SearchQuery(BaseModel):
    value: str | None = None
    regex: bool | None = None

class ColumnQuery(BaseModel):
    name: str | None = None
    searchable: bool | None = None
    orderable: bool | None = None
    search: SearchQuery | None = None

class TableQuery(BaseModel):
    draw: int
    start: int
    length: int
    columns: list[ColumnQuery]

app = FastAPI()


@app.get("/data")
def get_data(draw: int, start: int, length: int, columns: str = Query(...)):
    try:
        columns_list_dict = json.loads(columns)
        columns_list = [ColumnQuery(**col) for col in columns_list_dict]
        params = TableQuery(draw=draw, start=start, length=length, columns=columns_list)
    except json.JSONDecodeError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"JSON decode error: {e.msg}")
    except TypeError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"Type error: {e}")
    print(params)
    return params
```
You can check the results by calling the endpoint:

``
http://localhost:8000/data?draw=1&start=0&length=10&columns=[{"name":"column1","searchable":true,"orderable":true, "search": {"value": "some value", "regex": false}},{"name":"column2","searchable":false,"orderable":true}]
``

The idea is to send all columns as str that you will convert after to a JSON valid value. I don't think that's the best way, but if it works for you, you will need to change only the way you are sending the data.

**Reply by ldbolanos** [reply]
Thanks for the reply, Javier.

Although it's clever and hacky, I do not have that level of control over how the parameters are sent. I did find, however, the library allows to send a POST requests, and that allows me to easily parse the JSON body, list included.

**Comment by wangxin688** [comment]
comma sperated is not support natively by Fastapi, the best is use ?filed=value1&field=value2.
```python3
class QueryParams(BaseModel):
    filed: list[str] | None = Field(Query(default=[]))
```

**Comment by riwu** [comment]
Is there any plan to support this? Would like to have it too.
